### PR TITLE
Consolidate argument type of extraction function

### PR DESCRIPTION
**Pull request by navis**
Another proposal for extraction function. #2405 suggested to have multiple apply method for each type but this one is for making a single apply method with Object type parameter. Related to #2263.

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/column/ValueAccessor.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      return value != null ? String.valueOf(value) : nullToEmpty ? "" : null;
+    }
+  };
+
+  ValueAccessor COMPLEX = new AbstractValueAccessor()
+  {
+    @Override
+    public ValueType type()
+    {
+      return ValueType.COMPLEX;
```
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/column/ValueAccessor.java]: what is complex ?

File: processing/src/main/java/io/druid/query/extraction/ExtractionFn.java [hunk 2]
```diff
@@ -71,13 +74,9 @@
    */
   public String apply(Object value);
 
-  public String apply(String value);
```
Comment (b-slim) [inline processing/src/main/java/io/druid/query/extraction/ExtractionFn.java]: but this is backward incompatible ? it is totally a new API, can we have a proposal and maybe come with something compatible ?

File: processing/src/main/java/io/druid/query/extraction/AbstractExtractionFn.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * under the License.
+ */
+
+package io.druid.query.extraction;
+
+import io.druid.segment.column.ValueAccessor;
+
+/**
+ */
+public abstract class AbstractExtractionFn implements ExtractionFn
```
Comment (b-slim) [inline processing/src/main/java/io/druid/query/extraction/AbstractExtractionFn.java]: making it abstract class is somehow not very user friendly, assume i have my class that already extends another super class related to my own org, by making this abstract class you make it impossible to use it as an extraction function, unless i have to create the object within the class it self. i think we should find a way to make it as an interface.

File: processing/src/main/java/io/druid/segment/column/ValueAccessor.java (hunk 1 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/column/ValueAccessor.java]: It's not used, just made for completeness. Can be removed.

File: processing/src/main/java/io/druid/query/extraction/ExtractionFn.java (hunk 2 above)
Comment (navis) [inline processing/src/main/java/io/druid/query/extraction/ExtractionFn.java]: I've missed tagging `discuss` label on this. Yes, we should consider it but I've just wanted to show the intention.

File: processing/src/main/java/io/druid/query/extraction/AbstractExtractionFn.java (hunk 3 above)
Comment (navis) [inline processing/src/main/java/io/druid/query/extraction/AbstractExtractionFn.java]: It's an utility class for simplicity. We can always implement `ExtractionFn` directly.

**Pull request closed without merging**